dao/src/main/java/org/thingsboard/server/dao/sql/JpaAbstractDao.java

```diff
@@ -60,21 +85,30 @@ public D save(TenantId tenantId, D domain) {
         }
         setSearchText(entity);
         log.debug("Saving entity {}", entity);
-        if (entity.getUuid() == null) {
+        boolean isNew = entity.getUuid() == null;
+        if (isNew) {
             UUID uuid = Uuids.timeBased();
             entity.setUuid(uuid);
             entity.setCreatedTime(Uuids.unixTimestamp(uuid));
         }
-        entity = getRepository().save(entity);
-        return DaoUtil.getData(entity);
+
+        if (preSaveAction != null) {
+            preSaveAction.accept(entity);
+        }
+        if (TransactionSynchronizationManager.isActualTransactionActive()) {
+            return doSave(entity, isNew);
+        } else {
+            return transactionTemplate.execute(status -> doSave(entity, isNew));
+        }
     }
 
-    @Override
-    @Transactional
-    public D saveAndFlush(TenantId tenantId, D domain) {
-        D d = save(tenantId, domain);
-        getRepository().flush();
-        return d;
+    private D doSave(E entity, boolean isNew) {
+        if (isNew) {
+            entityManager.persist(entity);
+        } else {
+            entity = entityManager.merge(entity);
```

Comment:
as i understand, this is a replacement for repository.save()

Reply:
Yes

Reply:
People will asking why we replace the working repo.save() with something else